### ISSUE TITLE
Mockingboard speech interrupt occasionally delivered late

### ISSUE THREAD

**Issue by peterferrie** [AppleWin/AppleWin#958]
v1.30.1.0.  This might be specific to a debugger-interaction.  Very rarely when booting TR via "gg" from the debugger, the speech interrupt isn't delivered in time to be detected, so TR doesn't say "and it talks".  Then TR sets the IRQ handler to RTI, and some short time later the IRQ is delivered and then TR hangs.
I can't reproduce it on purpose at all.  I haven't noticed when exactly it happens (i.e., first boot, or after running AppleWin for a while and resetting).

**Comment by tomcw** [comment]
Thanks for raising this. I've seen this issue too, and like you can't get a solid reproducible.

NB. prior to 1.30.0 (ie. 1.29.x and earlier) the speech phonemes were played back asynchronously to the main emulation thread. The upshot of this was when you stepped a phoneme "kick" from within the AppleWin debugger, then the phoneme would complete immediately, ie. a ~20ms phoneme is comparable to the time to press SPACE to single-step. Anyway this "phoneme completing immediately" was also wrong, and would cause speech-code debugging issues too.

I'll take a deeper look.

**Comment by tomcw** [comment]
My repro:
- config: Phasor in slot-4 (or MB in slots 4 & 5)
- insert mb-audit (v0.6)
- From AppleWin logo screen: F7 (debugger), "gg"

This results in no speech chips being detected.

NB. using "g" (above) instead of "gg" allows all the speech chips to be detected.

**Comment by peterferrie** [comment]
Excellent.  Maybe in 'gg' mode the interrupt should be delivered
immediately since it makes no sense to wait?  Or that might introduce
its own set of issues when checking values later.

**Comment by tomcw** [comment]
I still want the IRQ to fire after the correct number of cycles for the phoneme.

The problem with "delivering the interrupt immediately" is this SSI263 code:

```
	lda		#CONTROL_H
	ldy		#SSI_CTTRAMP
	sta		(SSI263Base),y

	; Warning: stepping this won't yield an IRQ, since the phoneme completes before H->L (assuming prev mode=MODE_IRQ_DISABLED)
	lda		#MODE_PHONEME_TRANSITIONED_INFLECTION+$00	; DR1:DR0=3 , P5:P0=$00
	ldy		#SSI_DURPHON
	sta		(SSI263Base),y

	lda		#CONTROL_L				; H->L sets SSI263 operation mode = DR1:DR0, ie. IRQ active
	ldy		#SSI_CTTRAMP
	sta		(SSI263Base),y
```

ie, single-stepping at the "warning" comment above.

This was the old AppleWin problem before with with the previous asynchronous method (ie. a 2nd thread for phoneme playback).

Anyway your 'gg' problem should now be fixed at 0ca9729 - I was just missing a bit of state for the 'gg' (full-speed) path.

I'll try to do a release in the coming days.

**Comment by tomcw** [comment]
@peterferrie - try this [new build of AppleWin 1.30.2.0](https://github.com/AppleWin/AppleWin/releases/tag/v1.30.2.0) which fixes this issue.
